Dynamic series adjustment

yomek33 opened the pull request:
Add two new operation modes to the RunMetrics function:  `gradual-change` and `double-halve`:

## double-halve
usage
 `./avalanche --operation-mode=double-halve --series-change-interval=30 --series-count=20`

This mode alternately doubles and halves the series count at regular intervals.
The series count is doubled on one tick and halved on the next, ensuring it never drops below 1.

## gradual-change
usage
`./avalanche --operation-mode=gradual-change --series-change-interval=30 --series-change-rate=10 series-count=20`

This mode gradually changes the series count by a fixed rate at regular intervals
The series count is incremented by seriesChangeRate on each tick, ensuring it never drops below 1.

## Tests
- Added unit tests to cover the new double-halve and gradual-change modes.
- Verified that the series count adjusts correctly and does not fall below 1.

cstyan commented on cmd/avalanche.go:
```diff
@@ -30,25 +30,28 @@ import (
+	metricCount          = kingpin.Flag("metric-count", "Number of metrics to serve.").Default("500").Int()
+	labelCount           = kingpin.Flag("label-count", "Number of labels per-metric.").Default("10").Int()
+	seriesCount          = kingpin.Flag("series-count", "Number of series per-metric.").Default("10").Int()
+	seriesChangeRate     = kingpin.Flag("series-change-rate", "The rate at which the number of active series changes over time. Positive value increases the series, negative value decreases the series.").Default("10").Int()
```
for the flags that only apply to specific operation modes, can we specify which they work for in the help text?

yomek33 replied:
I added some text for that.
And I also tried to add some details for usage of modes like this🤔
![image](https://github.com/user-attachments/assets/273738b2-d594-4517-a47d-c85346041feb)

cstyan reviewed (commented):
Looking good! Just a few minor implementation comments.

Looks like the tests are failing, there's a make target `make test` which should run them the same way as Circle CI is where they're failing.

yomek33 commented on cmd/avalanche.go:
```diff
@@ -30,25 +30,28 @@ import (
+	labelInterval        = kingpin.Flag("series-interval", "Change series_id label values every {interval} seconds.").Default("60").Int()
+	metricInterval       = kingpin.Flag("metric-interval", "Change __name__ label values every {interval} seconds.").Default("120").Int()
+	seriesChangeInterval = kingpin.Flag("series-change-interval", "Change the number of series every {interval} seconds.").Default("10").Int()
+	operationMode        = kingpin.Flag("operation-mode", "Mode of operation: 'gradual-change', 'series-spike', 'double-halve'").Default("default").String()
```
I will add spike mode in next PR! So, I deleted this part of text in this PR.

cstyan replied:
:+1: sounds good

yomek33 commented on metrics/serve.go:
```diff
@@ -153,6 +154,59 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
+			}
+		}()
+	default:
+	}
```
I fixed🙆‍♀️

yomek33 commented on metrics/serve_test.go:
```diff
@@ -0,0 +1,139 @@
+		assert.Equal(t, expectedCount, updatedCount, "3 Decreased series count should be %d but got %d", expectedCount, updatedCount)
+	case <-time.After(updateNotifyTimeout):
+		t.Fatal("Did not receive update notification for series count doubling in time")
+	}
```
Thank you! For the gradual-change tests, I changed to utilized require.Eventually.

yomek33 commented on metrics/serve.go:
```diff
@@ -153,6 +154,59 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
+				default:
+				}
+			}
+		}()
```
I have made the changes!

yomek33 commented:
Additionally, I made the following changes:

Added a license to the project. This also resolved the CI issues. I set the license year to 2022 to match the other files, but please let me know if it should be updated to 2024.
Modified the logic for each Mode. Now, the SeriesCount is adjusted according to the Mode, and the correct SeriesCount is used where necessary.
Please review the updated PR and let me know if you have any further suggestions or concerns.
Thank you!

cstyan commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/prometheus-community/avalanche
 
-go 1.14
+go 1.17
```
we should update to the latest go 1.22 version

cstyan commented on cmd/avalanche.go:
```diff
@@ -30,36 +31,68 @@ import (
 func main() {
 	kingpin.Version(version.Print("avalanche"))
 	log.SetFlags(log.Ltime | log.Lshortfile) // Show file name and line in logs.
 	kingpin.CommandLine.Help = "avalanche - metrics test server"
+	kingpin.CommandLine.Help = "avalanche - metrics test server\n" +
```
I think we can delete line 63?

cstyan commented on cmd/avalanche.go:
```diff
@@ -30,36 +31,68 @@ import (
+	valueInterval        = kingpin.Flag("value-interval", "Change series values every {interval} seconds.").Default("30").Int()
+	labelInterval        = kingpin.Flag("series-interval", "Change series_id label values every {interval} seconds.").Default("60").Int()
+	metricInterval       = kingpin.Flag("metric-interval", "Change __name__ label values every {interval} seconds.").Default("120").Int()
+	seriesChangeInterval = kingpin.Flag("series-change-interval", "Change the number of series every {interval} seconds.").Default("10").Int()
```
should this one say that it applies to both the new operation modes?

cstyan commented on metrics/serve.go:
```diff
@@ -103,19 +103,112 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
 
+	var currentSeriesCount int
+
+	switch seriesOperationMode {
+	case "double-halve":
```
it might be worth breaking the body of each `case` plus the default into 3 separate functions that can be called

yomek33 replied:
I've made the changes as you suggested👀

cstyan commented on metrics/serve.go:
```diff
@@ -103,19 +103,112 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
+		}()
+	}
+
 	go func() {
```
we should think about whether to move the value tick and series ticks to their own functions as well, and call them inside each of the new operation mode cases in the switch statement _if_ they are needed

cstyan reviewed (commented):
we could have another PR to move the new operation modes to a separate sub command: https://pkg.go.dev/github.com/alecthomas/kingpin#readme-sub-commands

a few more comments about organization/structure of the code

cstyan commented:
If you rebase or merge in master it looks like Ben updated the build dependencies recently, so we don't need the change to upgrade to `go1.21` anymore. Also, for the DCO check to pass ([here](https://github.com/prometheus-community/avalanche/pull/64/checks?check_run_id=27632683230)) you need to sign commits with `git commit -s`.

yomek33 commented:
marged and I signed

cstyan commented on metrics/serve.go:
```diff
@@ -82,6 +82,135 @@ func cycleValues(labelKeys, labelValues []string, seriesCount, seriesCycle int)
+	registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+	cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+
+	go func() {
```
just for consistency, can we have this function and `handleGradualChangeMode` just do their loop, and move the `go func() { ... }` wrapper to where they are called in `RunMetrics`

yomek33 replied:
moved👍

cstyan commented on cmd/avalanche.go:
```diff
@@ -30,36 +31,67 @@ import (
+	seriesCount          = kingpin.Flag("series-count", "Number of series per-metric.").Default("10").Int()
+	seriesChangeRate     = kingpin.Flag("series-change-rate", "The rate at which the number of active series changes over time. Applies to 'gradual-change' mode.").Default("10").Int()
+	maxSeriesCount       = kingpin.Flag("max-series-count", "Maximum number of series to serve. Applies to 'gradual-change' mode.").Default("10000").Int()
+	minSeriesCount       = kingpin.Flag("min-series-count", "Minimum number of series to serve. Applies to 'gradual-change' mode.").Default("0").Int()
```
lets set this to something like 100 or 1000, 0 doesn't make a lot of sense practically

cstyan commented on metrics/serve.go:
```diff
@@ -103,62 +219,47 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
-			}
-		}
-	}()
+	var currentSeriesCount int
```
```suggestion
```

cstyan commented on metrics/serve.go:
```diff
@@ -103,62 +219,47 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
-	}()
+	switch seriesOperationMode {
+	case "double-halve":
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+		go handleDoubleHalveMode(metricCount, metricLength, metricCycle, seriesCycle, labelKeys, labelValues, seriesCount, changeSeriesTick, updateNotify)
+		go handleValueTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, valueTick)
+		go handleSeriesTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, seriesTick)
```
```suggestion
		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
		cycleValues(labelKeys, labelValues, seriesCount, seriesCycle)
		go handleDoubleHalveMode(metricCount, metricLength, metricCycle, seriesCycle, labelKeys, labelValues, seriesCount, changeSeriesTick, updateNotify)
		go handleValueTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, valueTick)
		go handleSeriesTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, seriesTick)
```

yomek33 replied:
I believe currentSeriesCount should be used instead of seriesCount when passing it to the functions handleValueTicks and handleSeriesTicks. 🤔
For example, during a gradual change, currentSeriesCount increases incrementally (e.g., 10, 20, 30, ...). If seriesCount (which might be a constant value like 500) is passed instead, it would disrupt the incremental increase and lead to inconsistencies in the series processing. This could result in an unexpected jump to seriesCount=500, instead of following the intended gradual progression.

cstyan commented on metrics/serve.go:
```diff
@@ -103,62 +219,47 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
-	}()
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, minSeriesCount, seriesCycle)
+		go handleGradualChangeMode(metricCount, metricLength, metricCycle, seriesCycle, labelKeys, labelValues, seriesChangeRate, maxSeriesCount, minSeriesCount, changeSeriesTick, updateNotify)
+		go handleValueTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, valueTick)
+		go handleSeriesTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, seriesTick)
```
```suggestion
		go handleGradualChangeMode(metricCount, metricLength, metricCycle, seriesCycle, labelKeys, labelValues, seriesChangeRate, maxSeriesCount, minSeriesCount, changeSeriesTick, updateNotify)
		go handleValueTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, valueTick)
		go handleSeriesTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, seriesTick)
```

cstyan commented on metrics/serve.go:
```diff
@@ -103,62 +219,47 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
+		go handleValueTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, valueTick)
+		go handleSeriesTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, seriesTick)
+
+	default:
+		currentSeriesCount = seriesCount
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, seriesCount, seriesCycle)
+		go handleValueTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, valueTick)
+		go handleSeriesTicks(&labelKeys, &labelValues, &currentSeriesCount, &seriesCycle, updateNotify, seriesTick)
+		go handleMetricTicks(&metricCount, &metricLength, &metricCycle, &labelKeys, updateNotify, metricTick)
```
```suggestion
default:
		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
		cycleValues(labelKeys, labelValues, seriesCount, seriesCycle)
		go handleValueTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, valueTick)
		go handleSeriesTicks(&labelKeys, &labelValues, &seriesCount, &seriesCycle, updateNotify, seriesTick)
		go handleMetricTicks(&metricCount, &metricLength, &metricCycle, &labelKeys, updateNotify, metricTick)
```

cstyan reviewed (commented):
I left a few comments for code change suggestions; these simplify the code a bit and also make sure that the two new operation modes have metrics available for scraping immediately. We either need this change _or_ to set `currentSeriesCount = seriesCount` in each of the two new modes as you've done in the default case.

cstyan commented on metrics/serve.go:
```diff
@@ -177,19 +177,19 @@ func handleDoubleHalveMode(metricCount, metricLength, metricCycle, seriesCycle i
 }
 
 func handleGradualChangeMode(metricCount, metricLength, metricCycle, seriesCycle int, labelKeys, labelValues []string, seriesChangeRate, maxSeriesCount, minSeriesCount int, changeSeriesTick *time.Ticker, updateNotify chan struct{}) {
```
left a comment on slack as well, but this function still needs to take in a pointer for the `seriesCount` so that when it modifies the value the other go routines know the correct value for `seriesCount`

cstyan commented:
@yomek33 good work!
